universal-application-tool-0.0.1/app/views/applicant/ApplicantProgramBlockEditView.java

```diff
@@ -199,34 +254,67 @@ private Tag renderReviewButton(Params params) {
   }
 
-  private Optional<Tag> renderSkipFileUploadButton(Params params) {
-    // Cannot skip required file questions.
-    if (!params.block().getQuestions().get(0).isOptional()) {
-      return Optional.empty();
-    }
-    String skipUrl =
-        routes.ApplicantProgramBlocksController.skipFile(
-                params.applicantId(), params.programId(), params.block().getId(), params.inReview())
-            .url();
-    return Optional.of(
-        a().attr(HREF, skipUrl)
-            .withText(params.messages().at(MessageKey.BUTTON_SKIP_FILEUPLOAD.getKeyName()))
-            .withId("skip-fileupload-button")
-            .withClasses(ApplicantStyles.BUTTON_REVIEW));
-  }
-
   private Tag renderNextButton(Params params) {
     return submitButton(params.messages().at(MessageKey.BUTTON_NEXT_BLOCK.getKeyName()))
         .withClasses(ApplicantStyles.BUTTON_BLOCK_NEXT)
         .withId("cf-block-submit");
   }
 
+  private Optional<Tag> renderContinueButton(Params params) {
+    if (!hasUploadedFile(params)) {
+      return Optional.empty();
+    }
+    Tag button =
+        submitButton(params.messages().at(MessageKey.BUTTON_KEEP_FILE.getKeyName()))
+            .attr(FORM, FILEUPLOAD_CONTINUE_FORM_ID)
+            .withClasses(ApplicantStyles.BUTTON_BLOCK_NEXT)
+            .withId("fileupload-continue-button");
+    return Optional.of(button);
+  }
+
+  private Optional<Tag> renderDeleteButton(Params params) {
+    String buttonText;
+    String buttonId;
+    if (hasUploadedFile(params)) {
+      buttonText = params.messages().at(MessageKey.BUTTON_DELETE_FILE.getKeyName());
+      buttonId = "fileupload-delete-button";
+    } else if (!hasRequiredQuestion(params)) {
+      buttonText = params.messages().at(MessageKey.BUTTON_SKIP_FILEUPLOAD.getKeyName());
+      buttonId = "fileupload-skip-button";
+    } else {
+      // Cannot skip required file questions.
+      return Optional.empty();
+    }
+    Tag button =
+        submitButton(buttonText)
+            .attr(FORM, FILEUPLOAD_DELETE_FORM_ID)
+            .withClasses(ApplicantStyles.BUTTON_REVIEW)
+            .withId(buttonId);
+    return Optional.of(button);
+  }
+
   private Tag renderUploadButton(Params params) {
+    String styles = ApplicantStyles.BUTTON_BLOCK_NEXT;
+    if (hasUploadedFile(params)) {
+      styles = ApplicantStyles.BUTTON_REVIEW;
+    }
     return submitButton(params.messages().at(MessageKey.BUTTON_UPLOAD.getKeyName()))
-        .withClasses(ApplicantStyles.BUTTON_BLOCK_NEXT)
+        .attr(FORM, BLOCK_FORM_ID)
+        .withClasses(styles)
         .withId("cf-block-submit");
   }
 
+  private boolean hasUploadedFile(Params params) {
+    return params.block().getQuestions().stream()
+        .map(ApplicantQuestion::createFileUploadQuestion)
+        .map(FileUploadQuestion::getFileKeyValue)
+        .anyMatch(maybeValue -> maybeValue.isPresent());
+  }
+
+  private boolean hasRequiredQuestion(Params params) {
```

Comment:
(non-blocking) I get it, but this function is confusing because its really only useful for file upload screens. Maybe if it was renamed "hasAtLeastOneRequriedQuestion" it would be a bit clearer when its used. I would probably opt for not making this a function if it is only used in one place.

Reply:
Renamed. I am a big fan of using short functions as a way to abstract lower-level details so the code is more concise and readable at a higher-level of abstraction, even when the functions are only used in one place. The added benefit is that it is usually very easy to look at a short function and visually verify if it is functionally correct because it should be doing one and only one simple thing.